validator/src/main/java/com/amazon/aoc/validators/CWMetricValidator.java

```diff
@@ -152,50 +153,57 @@ private void addMetrics(
    *
-   * @param toBeCheckedMetricList toBeCheckedMetricList
-   * @param baseMetricList baseMetricList
+   * @param expectedMetricList expectedMetricList
+   * @param actualMetricList actualMetricList
    */
-  private void compareMetricLists(List<Metric> toBeCheckedMetricList, List<Metric> baseMetricList)
+  private void compareMetricLists(List<Metric> expectedMetricList, List<Metric> actualMetricList)
       throws BaseException {
 
-    // load metrics into a hash set
-    Set<Metric> metricSet =
-        new TreeSet<>(
-            (Metric o1, Metric o2) -> {
-              // check namespace
-              if (!o1.getNamespace().equals(o2.getNamespace())) {
-                return o1.getNamespace().compareTo(o2.getNamespace());
-              }
-
-              // check metric name
-              if (!o1.getMetricName().equals(o2.getMetricName())) {
-                return o1.getMetricName().compareTo(o2.getMetricName());
-              }
-
-              // sort and check dimensions
-              List<Dimension> dimensionList1 = o1.getDimensions();
-              List<Dimension> dimensionList2 = o2.getDimensions();
-
-              // sort
-              dimensionList1.sort(Comparator.comparing(Dimension::getName));
-              dimensionList2.sort(Comparator.comparing(Dimension::getName));
-
-              return dimensionList1.toString().compareTo(dimensionList2.toString());
-            });
-    for (Metric metric : baseMetricList) {
-      metricSet.add(metric);
-    }
-    for (Metric metric : toBeCheckedMetricList) {
-      if (!metricSet.contains(metric)) {
-        throw new BaseException(
-            ExceptionCode.EXPECTED_METRIC_NOT_FOUND,
-            String.format(
-                "metric in %ntoBeCheckedMetricList: %s is not found in %nbaseMetricList: %s %n",
-                metric, metricSet));
+      Set<Metric> matchAny = new HashSet<>();
+      Set<Metric> matchExact = new HashSet<>();
+      for (Metric metric : expectedMetricList) {
+          metric.getDimensions().sort(Comparator.comparing(Dimension::getName));
+          if (metric.getDimensions().stream().anyMatch(d -> ANY_VALUE.equals(d.getValue()))) {
+              matchAny.add(metric);
+          } else {
+              matchExact.add(metric);
+          }
       }
-    }
+
+      Set<Metric> actualMetricSet = new HashSet<>();
+      for (Metric metric : actualMetricList) {
+         metric.getDimensions().sort(Comparator.comparing(Dimension::getName));
+         actualMetricSet.add(metric);
+      }
+      matchExact.removeAll(actualMetricSet);
+      if (!matchExact.isEmpty())  {
+          throw new BaseException(
+                  ExceptionCode.EXPECTED_METRIC_NOT_FOUND,
+                  String.format(
+                          "metric in %ntoBeCheckedMetricList: %s is not found in %nbaseMetricList: %s %n",
+                          matchExact.stream().findAny().get(), actualMetricSet));
```

Comment:
This will print only one of potentially many missing metrics, right? It also doesn't wait until the `matchAny` check is done, could we collect all the metrics we find missing and list them for better understanding of what exactly went wrong?

Reply:
Yes, this is how it works today.

We are able to identify all missing metrics because the deduplicated `actualMetrics` are all printed out. Since we know the template to compare, we can figure out the diff.